tests/shared/src/test/scala/scala/meta/tests/contrib/AssociatedCommentsTest.scala

```diff
@@ -39,19 +38,234 @@ class AssociatedCommentsTest extends FunSuite {
+
+    val defnTrait = input.find(_.is[Defn.Trait]).get
+
+    assertExpectations(input)(
+      leading = Map(
+        defnTrait -> Set("/** leading 1 */", "/* leading 2 */", "// leading 3")
+      )
+    )
+  }
+
+  ignore("single trailing comment at the end of a file") {
+    val input =
+      """|object A {
+         |} // trailing
+         |""".stripMargin.parse[Source].get
+
+    val defnObject = input.find(_.is[Defn.Object]).get
+    val template = input.find(_.is[Template]).get // overlaps with `object`
+
+    assertExpectations(input)(
+      trailing = Map(
+        defnObject -> Set("// trailing"),
+        template -> Set("// trailing")
+      )
+    )
+  }
+
+  test("multiple trailing comments in a single line at the end of a file") {
+    val input =
+      """|class A {
+         |} /** trailing 1 */ /* trailing 2 */ // trailing 3
+         |""".stripMargin.parse[Source].get
+
+    val defnClass = input.find(_.is[Defn.Class]).get
+    val template = input.find(_.is[Template]).get // overlaps with `class`
+
+    val expectedComments = Set("/** trailing 1 */", "/* trailing 2 */", "// trailing 3")
+    assertExpectations(input)(
+      trailing = Map(
+        defnClass -> expectedComments,
+        template -> expectedComments
+      )
+    )
+  }
+
+  test("multiple trailing comments in different lines at the end of a file") {
+    val input =
+      """|trait A {
+         |} /** trailing 1 */
+         |/* trailing 2 */
+         |// trailing 3
+         |""".stripMargin.parse[Source].get
+
+    val defnTrait = input.find(_.is[Defn.Trait]).get
+    val template = input.find(_.is[Template]).get // overlaps with `trait`
+
+    val expectedComments = Set("/** trailing 1 */", "/* trailing 2 */", "// trailing 3")
+    assertExpectations(input)(
+      trailing = Map(
+        defnTrait -> expectedComments,
+        template -> expectedComments
+      )
+    )
+  }
+
+  test("comment not associated to any tree") {
+    val input =
+      """|object A {
+         | // foo
+         |}
+         |""".stripMargin.parse[Source].get
+
+    assertExpectations(input)(
+      leading = Map.empty,
+      trailing = Map.empty
+    )
+  }
+
+  test("tree with both leading and trailing comments") {
+    val input =
+      """|object A {
+         | // leading
+         | val x = 0 // trailing
+         | var y = false
+         |}
+         |""".stripMargin.parse[Source].get
+
+    val defnVal = input.find(_.is[Defn.Val]).get
+    val litInt = input.find(_.is[Lit.Int]).get // overlaps with `val`
+
+    assertExpectations(input)(
+      leading = Map(
+        defnVal -> Set("// leading")
+      ),
+      trailing = Map(
+        defnVal -> Set("// trailing"),
+        litInt -> Set("// trailing")
+      )
+    )
+  }
+
+  test("multiple comments interleaved with trees in a single line") {
+    val input =
+      """|object A {
+         | /* comment 1 */ val /* comment 2 */ foo = /* comment 3 */ 0 /* comment 4 */
```

Comment:
What happens to comments 2 and 3? I would expect comment 2 to be leading for `foo` and comment 3 leading for `0`. Looks like a bug unrelated to this PR.

Reply:
@olafurpg that's something that got me confused too. It turns out that the parsing logic associates comments to **tokens**. If you inspect the internal state you will see that comments 2 and 3 are there:

```
AssociatedComments(
  Leading =
    val [28..31) => List(/*∙comment∙1∙*/)

  Trailing =
    val [28..31) => List(/*∙comment∙2∙*/)
    = [52..53) => List(/*∙comment∙3∙*/)
    0 [70..71) => List(/*∙comment∙4∙*/)
)
```

However, the `AssociatedCommentsTest` public api let you query comments associated to **trees**. The `leading` and `trailing` methods only search for comments associated to the first and last tokens (respectively) of the given tree node. Comments 2 and 3 are neither associated to one or the other of any tree. Not sure if this is a bug though. Wdyt?

Reply:
I think this is a bug, good catch. I opened https://github.com/scalameta/scalameta/issues/1539 to track it.